### PR TITLE
Add hw 1

### PR DESCRIPTION
**Pull request by GalReg**

### CODE REVIEW

**Comment by TikhobaevaOlga** [comment]
Добрый день! 
Спасибо за проделанную работу, в целом все отлично. Отдельно хочу отметить, что ноутбук очень структурированный, было приятно проверять такую работу!

**Ниже оставлю баллы и несколько комментариев по пунктам:**

1. Подготовьте данные к обучению - **0.5 балла**

Все пункты выполнены, но StandardScaler применен неверно. При нормализации нужно «обучить/настроить» и применить скелер к train выборке (fit_transform). К валидационной и тестовой выборкам скелер нужно уже только применять (transform), так как нужно сохранить масштаб, соответствующий обучающим данным. Это важный момент, который влияет на дальнейшие результаты!

2. Реализуйте класс бейзлайновой нейросетевой модели (MLP) для решения задачи - **2 балла**

Все пункты задания выполнены, есть аргументация выбора функции активации и кол-ва слоев. Хотя с последним в целом тоже можно было поэкспериментировать немного.

3. Напишите код цикла обучения - train-loop и валидации - eval-loop. В процессе обучения сохраняйте и визуализируйте на графике динамику функции потерь на тренировочной и валидационной выборках - **1 балл**

4. Обучите модель и проверьте ее качество - **1 балл**

Все пункты задания выполнены, есть эксперименты и выводы. Однако, кажется, что при прочих установленных гиперпараметрах выбор 200 эпох не совсем оптимален. В данной ситуации можно сделать несколько наблюдений о процессе обучения:

1. Поведение loss: На 50 эпохах loss стабильно снижается, что указывает на то, что модель всё ещё учится и не страдает от переобучения. Примерно на 60-й эпохе, судя по графику, наблюдается резкое увеличение loss, его начинает «болтать» на одном месте, что может указывать на нестабильное обучение или переобучение. Да, к 200 эпохе loss выравнивается, но становится даже больше, чем на 50 эпохах.

2. Метрика f1: Значение f1 на тестовых данных увеличивается лишь незначительно с 0.743 до 0.747, что говорит о том, что улучшение незначительное.

Учитывая эти моменты, оптимально выбирать 50 эпох, чтобы не затягивать обучение, избежать переобучения и сохранить качество модели на уровне. Или можно было попробовать еще промежуточные варианты, например, 60 или 120.

5. Улучшите архитектуру модели из пункта 3 и добейтесь увеличения качества - **3 балла**

Все пункты задания выполнены, есть аргументация, эксперименты и выводы. Получена модель, показавшая себя лучше бейзлайна. Можно было поэкспериментировать с размером батча, возможно, батч 32 маловат для BatchNorm, да и в целом можно было попробовать батч побольше и в прочих экспериментах. Плюс можно было попробовать не добавлять Dropout ко всем слоям, как это было в примере на занятии.

6. Принимаемые решения обоснованы (почему выбрана определенная архитектура/гиперпараметр/оптимизатор/проеобразование и т.п.) - **1 балл**
7. Обеспечена воспроизводимость решения: зафиксированы random_state, ноутбук воспроизводится от начала до конца без ошибок - **1 балл**

**Итоговый балл: 9.5**

**Review by TikhobaevaOlga — APPROVED** [review]

**Review by pacifikus — APPROVED** [review]
